### ISSUE TITLE
[BUG][CRASH] DisplayBuffer atoms being outside window range causes various crashes

### ISSUE THREAD

**Issue by Calum-J-I** [mawww/kakoune#4674]
### Version of Kakoune

v2021.11.08-309-g9ebd0cd9

### Reproducer

```bash
git clone https://github.com/mawww/kakoune.git
git checkout 9ebd0cd9c12523e02dae96130454791e48479704  # master at the time of writing
make # or make debug=yes to fail in asserts rather than syscalls
kak -n
# in the scratch buffer add a whole bunch of tabs till the screen scrolls horizontally and then execute
# add-highlighter window/ show-whitespaces
```

### Outcome

editor crashes for various reasons. most common is a negative value for tab length which fails an assert on casting to size_t. if you don't build in debug mode then that just causes a bad alloc.

inspecting the process it seems that the display atoms include off screen tabs which break ShowWhitespaces tab rendering logic. unsure if this is specific to tabs or they're just an easy way to break highlighter code.

### Expectations

tab characters are correctly drawn on the screen.

**Comment by eraserhd** [comment]
This was happening to me a lot last week (I thought it was related to kak-lsp), and I solved for now by backing out to commit `b5e565bd`, which works.  I think that makes `df79d0c2` the first bad commit?

**Comment by Screwtapello** [comment]
The display code was refactored recently and has already had a few bugfixes, in the following commits:

- df79d0c2453991c4762c6c6ffbccd714312c87f9
- 50cacc0758ea7e6acd7dcc350f8dbb2e6559a30e
- 195fe8fd29a1824372ff69e6ef8b8e5be8bc9a07
- c7fbf1f3905f77f136e898479352e590787434d2
